pkgs/native_assets_cli/lib/src/api/build_output.dart

```diff
@@ -4,48 +4,113 @@
-import 'dependencies.dart';
-import 'metadata.dart';
+import 'build_config.dart';
+import 'os.dart';
 
-abstract class BuildOutput {
-  /// Time the build this output belongs to started.
+part '../model/build_output.dart';
+
+/// The output of a `build.dart` invocation.
+///
+/// A package can have a toplevel `build.dart` script. If such a script exists,
+/// it will be automatically run, by the Flutter and Dart SDK tools. The script
+/// is expect to produce a specific output which [BuildOutput] can produce.
+abstract final class BuildOutput {
+  /// Start time for the build of this output.
   ///
-  /// Rounded down to whole seconds, because [File.lastModified] is rounded
-  /// to whole seconds and caching logic compares these timestamps.
+  /// The [timestamp] is rounded down to whole seconds, because
+  /// [File.lastModified] is rounded to whole seconds and caching logic compares
+  /// these timestamps.
   DateTime get timestamp;
-  List<Asset> get assets;
-  Dependencies get dependencies;
-  Metadata get metadata;
 
+  /// The assets produced by this build.
+  ///
+  /// In dry runs, the assets for all [Architecture]s for the [OS] specified in
+  /// the dry run must be provided.
+  Iterable<Asset> get assets;
+
+  /// The files used by this build.
+  ///
+  /// If any of the files in [dependencies] are modified after [timestamp], the
+  /// build will be re-run.
+  Iterable<Uri> get dependencies;
+
+  /// Create a build output.
+  ///
+  /// The [timestamp] must be before any [dependencies] are read by the build
+  /// this output belongs to. If the [BuildOutput] object is created at the
+  /// beginning of the `build.dart` script, [timestamp] can be omitted and will
+  /// default to [DateTime.now]. The [timestamp] is rounded down to whole
+  /// seconds, because [File.lastModified] is rounded to whole seconds and
+  /// caching logic compares these timestamps.
+  ///
+  /// The [Asset]s produced by this build or dry-run can be provided to the
+  /// constructor as [assets], or can be added later using [addAsset] and
+  /// [addAssets]. In dry runs, the [Architecture] for [CCodeAsset]s can be
+  /// omitted.
+  ///
+  /// The files used by this build must be provided to the constructor as
+  /// [dependencies], or can be added later with [addDependency] and
+  /// [addDependencies]. If any of these files are modified after [timestamp],
+  /// the build will be re-run. Typically these dependencies contain the
+  /// `build.dart` script itself, and the source files used in the build.
+  ///
+  /// Metadata can be passed to `build.dart` invocations of dependent packages.
+  /// It must be provided to the constructor as [metadata], or added later with
+  /// [addMetadatum] and [addMetadata].
   factory BuildOutput({
     DateTime? timestamp,
-    List<Asset>? assets,
-    Dependencies? dependencies,
-    Metadata? metadata,
+    Iterable<Asset>? assets,
+    Iterable<Uri>? dependencies,
+    Map<String, Object>? metadata,
```

Comment:
This is not particularly extensible.

Imagine we make `flutter build` add some flutter specific things to `BuildConfig` and `BuildOutput`. That requires us to add flutter specifics here. But command line dart users that only use `dart build` don't need flutter specifics.

Reply:
`build.dart` scripts do not know if they are being invoked from Flutter or not. (They might be able to infer it from the targetOS being Android, but not if it's MacOS.) So they can't just start outputting more data.
Flutter and Dart (and other launchers/embedders) have to agree on the protocol.

Within agreement on the protocol we can have some well-defined extension points. For example as you suggested having a list of accepted asset types and then arbitrary assets in the list of assets.
But Flutter can't add something to the protocol that the other embedders don't know about, a `build.dart` writer would have no way of targetting it.

We can only start supporting something like that if we add a `BuildConfig.sdk` -> that returns `'Dart'` or `'Flutter'`.
However, I'm not sure that that is a good idea.
The whole idea of native assets is that the package work both in Dart and Flutter. I don't want package authors to start branching on `BuildConfig.sdk` and not support one of the two SDKs.